File: k9mail/src/main/java/com/fsck/k9/message/html/HtmlProcessor.java
package com.fsck.k9.message.html;


import org.jsoup.nodes.Document;


public class HtmlProcessor {
    private final HtmlSanitizer htmlSanitizer;


    public static HtmlProcessor newInstance() {
        HtmlSanitizer htmlSanitizer = new HtmlSanitizer();
        return new HtmlProcessor(htmlSanitizer);
    }

    private HtmlProcessor(HtmlSanitizer htmlSanitizer) {
        this.htmlSanitizer = htmlSanitizer;
    }

    public String processForDisplay(String html) {
        Document document = htmlSanitizer.sanitize(html);
        addCustomHeadContents(document);

        return toCompactString(document);
    }

    private void addCustomHeadContents(Document document) {
        document.head().append("<meta name=\"viewport\" content=\"width=device-width\"/>" +
                HtmlConverter.cssStyleTheme() +
                HtmlConverter.cssStylePre());
    }

    public static String toCompactString(Document document) {
        document.outputSettings()
                .prettyPrint(false)
                .indentAmount(0);

        return document.html();
    }
}

File: library/html-cleaner/src/main/kotlin/app/k9mail/html/cleaner/HtmlProcessor.kt
package app.k9mail.html.cleaner

import org.jsoup.nodes.Document

class HtmlProcessor(
    private val customClasses: Set<String>,
    private val htmlHeadProvider: HtmlHeadProvider,
) {
    private val htmlSanitizer = HtmlSanitizer()

    fun processForDisplay(html: String): String {
        return htmlSanitizer.sanitize(html)
            .addCustomHeadContents()
            .addCustomClasses()
            .toCompactString()
    }

    private fun Document.addCustomHeadContents() = apply {
        head().append(htmlHeadProvider.headHtml)
    }

    private fun Document.toCompactString(): String {
        outputSettings()
            .prettyPrint(false)
            .indentAmount(0)

        return html()
    }

    private fun Document.addCustomClasses() = apply {
        if (customClasses.isNotEmpty()) {
            body().apply {
                customClasses.forEach(::addClass)
            }
        }
    }
}
